mypy/test/teststubtest.py

```diff
@@ -1177,6 +1180,142 @@ class C: ...
+                @final
+                def ham(self, obj: int) -> int: ...
+            """,
+            runtime="""
+            class D:
+                @final
+                def foo(self): pass
+                @final
+                @staticmethod
+                def bar(): pass
+                @final
+                @classmethod
+                def baz(cls): pass
+                @property
+                @final
+                def eggs(self): return 42
+                @final
+                @functools.lru_cache()
+                def ham(self, obj): return obj * 2
+            """,
+            error=None,
+        )
+        # Stub methods are allowed to have @final even if the runtime doesn't...
+        yield Case(
+            stub="""
+            class E:
+                @final
+                def foo(self) -> None: ...
+                @final
+                @staticmethod
+                def bar() -> None: ...
+                @final
+                @classmethod
+                def baz(cls) -> None: ...
+                @property
+                @final
+                def eggs(self) -> int: ...
+                @final
+                def ham(self, obj: int) -> int: ...
+            """,
+            runtime="""
+            class E:
+                def foo(self): pass
+                @staticmethod
+                def bar(): pass
+                @classmethod
+                def baz(cls): pass
+                @property
+                def eggs(self): return 42
+                @functools.lru_cache()
+                def ham(self, obj): return obj * 2
+            """,
+            error=None,
+        )
+        # ...But if the runtime has @final, the stub must have it as well
+        yield Case(
+            stub="""
+            class F:
+                def foo(self) -> None: ...
+            """,
+            runtime="""
+            class F:
+                @final
+                def foo(self): pass
+            """,
+            error="F.foo",
+        )
+        yield Case(
+            stub="""
+            class G:
+                @staticmethod
+                def foo() -> None: ...
+            """,
+            runtime="""
+            class G:
+                @final
+                @staticmethod
+                def foo(): pass
+            """,
+            error="G.foo",
+        )
+        yield Case(
+            stub="""
+            class H:
+                @classmethod
+                def foo(cls) -> None: ...
+            """,
+            runtime="""
+            class H:
+                @final
+                @classmethod
+                def foo(cls): pass
+            """,
+            error="H.foo",
+        )
+        yield Case(
+            stub="""
+            class I:
+                @property
+                def foo(self) -> int: ...
+            """,
+            runtime="""
+            class I:
+                @property
+                @final
```

Comment:
Just to be clear: `@final @property` won't generate an error, because `__final__` won't be set on a property object.

I think we can add a test case for it with a `XXX` note.